### ISSUE TITLE
Stockfish on Linux does not communicate with a GUI (if its Linux binary file is not launched manually first)

### ISSUE THREAD

**Issue by PedanticHacker** [official-stockfish/Stockfish#3683]
I'm creating a cross-platform chess GUI application that has Stockfish as its default chess engine.

I've noticed that Stockfish does not communicate with my chess GUI application on any Linux-based operating system (Manjaro, Ubuntu, Fedora, and others) if its Linux binary file is not double-clicked and launched first. After that, everything works normally.

The issue is not present on the Windows operating system, only on a Linux-based one. (Not sure about macOS, because I don't have access to it.)

I'm using an engine-communicating API of python-chess (the one that Niklas Fiekas [@niklasf on GitHub] created). This API works properly, so I'm guessing it's either Stockfish that has issues on Linux, or Linux itself is the culprit.

I just wanna get to the bottom of this, and solve the issue. I don't want Linux users to go manually launching the Stockfish's Linux binary file first, and only then play chess against Stockfish in my chess GUI application. That would be dumb.

What do you guys reckon is the culprit here?

**Comment by NightlyKing** [comment]
It's not an issue. It's a security measure that Windows simply lacks. Either the user starts SF manually or the program changes the permissions which requires a password input from the user. For further details take a look at this: https://www.linux.com/training-tutorials/understanding-linux-file-permissions/

**Comment by PedanticHacker** [comment]
@NightlyKing, thanks for that link. So how exactly should I run the `chmod` command?

**Comment by joergoster** [comment]
@PedanticHacker `chmod +x stockfish`

Or the GUI method: right-click on the file (stockfish binary you want to run), select the _Permissions_ tab, tick checkbox _Allow executing file as program_

**Comment by PedanticHacker** [comment]
@joergoster, thank you for this wonderful tip! I have solved my issue now.

The code of my chess GUI application runs on Python, so I just used Python's `os.chmod()` function, and passsed to it all the needed things (firstly, finding out that Python provides a `chmod()` function, and then immediately reading its documentation).

I have tested, and yes, Stockfish runs on Linux immediately now, because the Stockfish binary file is marked as an executable file (the *Permissions* tab has the checkbox *Is executable* checked).

Does this apply to macOS as well? Will I need to use this `os.chmod()` function for the Stockfish macOS binary file as well? Not sure, but I'm guessing macOS also has modes for files like Linux, yes?

**Comment by dsmsgms** [comment]
Very strange that the file didn't have executable permissions. I have used distro-built or manually compiled versions of Stockfish and never had such problem with GUIs.

**Comment by PedanticHacker** [comment]
@dsmsgms, well, I'm creating my own chess GUI. The chess GUI that you've used had probably changed the Stockfish binary file's executable permission behind the scenes without you even knowing it.

But now I resolved this in my own custom GUI, and Stockfish now works in my GUI on Linux as well (as it already did on Windows).

If anyone has macOS, please tell me how macOS is handling file permissions.

**Comment by dsmsgms** [comment]
Just compiled from scratch:
```
$ ls -l
-rwxr-xr-x 1 douglas wheel 47523320 Aug 28 17:21 stockfish
```

macOS is literally Unix, the file permissions work the same way.

**Comment by PedanticHacker** [comment]
@dsmsgms, ah, so I need to do the same for macOS as I did for Linux. Thanks.

Can you please upload that binary somewhere. I will, of course, acknowledge you as the Stockfish binary provider in my chess GUI application.

Uhm, does your macOS binary include the NNUE as well? 🤔

**Comment by dsmsgms** [comment]
I don't have macOS

**Comment by nguyenpham** [comment]
I am a macOS user. To install Stockfish I have either brew it or build from source code. All are runnable without chmod. NNUE is embedded already in those binaries.

**Comment by PedanticHacker** [comment]
Is it true that Linux builds (binaries) of Stockfish can also be run under macOS?

**Comment by nguyenpham** [comment]
No! If Linux binaries can run on macOS, our life become much easier ;)
BTW, macOS is not Linux but Unix, quite a different OS. I have also tried running some Stockfish binaries for Linux on macOS. All just print out only one line: "exec format error"

**Comment by PedanticHacker** [comment]
@nguyenpham, ah, I see.

Can anyone compile a macOS binary from the latest Stockfish source for me? Please make it x86_64-avx2, if possible.

**Comment by ppigazzini** [comment]
@PedanticHacker the wiki has the instructions to build Stockfish on macOS:
https://github.com/glinscott/fishtest/wiki/Building-Stockfish-on-macOS

ps: not a macOS user, hoping that the page is up to date

**Comment by ArminHHJ** [comment]
SF-14-270821-3349-x86-64-bmi2-clang-profil-aj.zip
https://filehorst.de/d/dGwmtvyb

**Comment by PedanticHacker** [comment]
I would gladly compile my own Stockfish binary file for macOS, but I don't have a Mac, or a MacBook.

@ArminHHJ, thanks for that link, I really appreciate it! It's a German site, I would never have found this.

Anyway, all Stockfish binary files for Linux that are available on https://abrok.eu/stockfish have the executable permission disabled. Maybe it's because of some potential security risk. I don't know.

**Comment by ppigazzini** [comment]
> Anyway, all Stockfish binary files for Linux that are available on https://abrok.eu/stockfish have the executable permission disabled. Maybe it's because of some potential security risk. I don't know.

I downloaded (in linux) one of abrok zip and stockfish has the executable permission enabled.
```bash
wget https://abrok.eu/stockfish/builds/ad357e147a1b8481a04761d726ce1db14115a68f/linux64modern/stockfish_21082721_x64_modern.zip
unzip stockfish_21082721_x64_modern.zip
ls -la
 total 74916
drwxr-xr-x  2 usr00 usr00     4096 Aug 29 22:39 .
drwxr-xr-x 24 usr00 usr00     4096 Aug 29 22:39 ..
-rw-r--r--  1 usr00 usr00     4669 Aug 27 21:50 AUTHORS
-rw-r--r--  1 usr00 usr00    35821 Aug 27 21:50 Copying.txt
-rw-r--r--  1 usr00 usr00      232 Aug 27 21:50 info.txt
-rwxr-xr-x  1 usr00 usr00 48307144 Aug 27 21:50 stockfish_21082721_x64_modern
-rw-r--r--  1 usr00 usr00 28345942 Aug 27 21:50 stockfish_21082721_x64_modern.zip
```

**Comment by PedanticHacker** [comment]
@ppigazzini, I have noticed the strangest thing.

I booted into Manjaro Linux from my USB key and downloaded a Stockfish binary file for Linux from https://abrok.eu/stockfish (while on Manjaro Linux, of course). Then I inspected that file (on Manjaro Linux) and noticed that, yes, it **is** executable.

However, then I booted back into Windows and downloaded a Stockfish binary file for Linux (from within Windows), then copied that file on a USB key, booted into Manjaro Linux again, and inspected the file (that was downloaded from within Windows) on Manjaro Linux, and the file was shown as **not** executable.

Isn't that strange?

**Comment by PedanticHacker** [comment]
I have learned that if you want to have an executable Stockfish binary file for Linux, you **must** download it on a Linux-based operating system (like Manjaro, or Ubuntu, or Fedora, etc.), because if you download a Stockfish binary file for Linux from within Windows, then that binary file will **not** be executable and will then **not** work on Linux as expected.

I am creating my chess GUI application from within Windows, and I got burned by the fact that Windows screws up any executable Linux file that you download from within Windows.

I have learned my lesson. And this message is for anyone that might experience the same frustration as I did.
